tests/python/pants_test/test_base.py

```diff
@@ -0,0 +1,559 @@
+                            as its first parameter for instance method binding.
+    """
+    assert not hasattr(cls, method_name), (
+      'a test with name `{}` already exists on `{}`!'.format(method_name, cls.__name__)
+    )
+    assert method_name.startswith('test_'), '{} is not a valid test name!'.format(method_name)
+    setattr(cls, method_name, method)
+
+
+class TestBase(unittest.TestCase):
+  """A baseclass useful for tests requiring a temporary buildroot.
+
+  :API: public
+  """
+
+  def build_path(self, relpath):
+    """Returns the canonical BUILD file path for the given relative build path.
+
+    :API: public
+    """
+    if os.path.basename(relpath).startswith('BUILD'):
+      return relpath
+    else:
+      return os.path.join(relpath, 'BUILD')
+
+  def create_dir(self, relpath):
+    """Creates a directory under the buildroot.
+
+    :API: public
+
+    relpath: The relative path to the directory from the build root.
+    """
+    path = os.path.join(self.build_root, relpath)
+    safe_mkdir(path)
+    return path
+
+  def create_workdir_dir(self, relpath):
+    """Creates a directory under the work directory.
+
+    :API: public
+
+    relpath: The relative path to the directory from the work directory.
+    """
+    path = os.path.join(self.pants_workdir, relpath)
+    safe_mkdir(path)
+    return path
+
+  def _collect_invalid_for(self, relpath):
+    """Before touching the given path (relative to the build root) computes all affected paths.
+
+    Many python operations implicitly create parent directories, so we assume that touching a
+    file located below directories that do not currently exist will result in their creation.
+
+    This emulates (to some degree) the types of events we'd expect to see from watchman when a
+    user creates or updates a file. A created/removed file/directory results in two events: one
+    for the parent, and one for the child. An updated file on the other hand only sends a single
+    event (for the child and not its parent).
+    """
+    # Invalidate for any path components which will be directly or indirectly created. The first
+    # component is always considered to have been invalidated.
+    transitively_invalidated = True
+    for p in recursive_dirname(relpath):
+      directly_invalidated = not os.path.exists(os.path.join(self.build_root, p))
+      if directly_invalidated or transitively_invalidated:
+        yield p
+      transitively_invalidated = directly_invalidated
+
+  def create_file(self, relpath, contents='', mode='wb'):
+    """Writes to a file under the buildroot.
+
+    :API: public
+
+    relpath:  The relative path to the file from the build root.
+    contents: A string containing the contents of the file - '' by default..
+    mode:     The mode to write to the file in - over-write by default.
+    """
+    if self._graph_helper is not None:
+      self._graph_helper.scheduler.invalidate_files(list(self._collect_invalid_for(relpath)))
```

Comment:
The invalidation should probably happen after the write, rather than before. Otherwise if we end up with e.g. background threads consuming these, they may consume the events too quickly.

Reply:
Not expecting any threads here, but sure... can't hurt.